services/venice-server/src/main/java/com/linkedin/venice/stats/ServerReadQuotaUsageStats.java

```diff
@@ -0,0 +1,182 @@
+ *      report rejections at the store level, regardless if it came from current or backup version due to stale routing.
+ *   2. VeniceVersionedStats is tracking current and future. However, for the purpose of quota we only care about
+ *      current and backup.
+ */
+public class ServerReadQuotaUsageStats extends AbstractVeniceStats {
+  private final Sensor currentRequestedQPS; // requested query per second for current version
+  private final Sensor backupRequestedQPS; // requested query per second for backup version
+  private final Sensor currentRequestedKPS; // requested key per second for current version
+  private final Sensor backupRequestedKPS; // requested key per second for backup version
+  private final Sensor rejectedQPS; // rejected query per second
+  private final Sensor rejectedKPS; // rejected key per second
+  private final Sensor allowedUnintentionallyKPS; // allowed KPS unintentionally due to error or insufficient info
+  private final Sensor usageRatioSensor; // requested kps divided by nodes quota responsibility
+  private final Int2ObjectMap<ServerReadQuotaVersionedStats> versionedStats = new Int2ObjectOpenHashMap<>();
+  private final AtomicInteger currentVersion = new AtomicInteger(0);
+  private final AtomicInteger backupVersion = new AtomicInteger(0);
+  private final Time time;
+  private final MetricConfig metricConfig;
+
+  public ServerReadQuotaUsageStats(MetricsRepository metricsRepository, String name) {
+    this(metricsRepository, name, new SystemTime());
+  }
+
+  public ServerReadQuotaUsageStats(MetricsRepository metricsRepository, String name, Time time) {
+    super(metricsRepository, name);
+    currentRequestedQPS = registerSensor(
+        new AsyncGauge((ignored, ignored2) -> getVersionedRequestedQPS(currentVersion.get()), "current_quota_request"));
+    backupRequestedQPS = registerSensor(
+        new AsyncGauge((ignored, ignored2) -> getVersionedRequestedQPS(backupVersion.get()), "backup_quota_request"));
+    currentRequestedKPS = registerSensor(
+        new AsyncGauge(
+            (ignored, ignored2) -> getVersionedRequestedKPS(currentVersion.get()),
+            "current_quota_request_key_count"));
+    backupRequestedKPS = registerSensor(
+        new AsyncGauge(
+            (ignored, ignored2) -> getVersionedRequestedKPS(backupVersion.get()),
+            "backup_quota_request_key_count"));
+    rejectedQPS = registerSensor("quota_rejected_request", new Rate());
+    rejectedKPS = registerSensor("quota_rejected_key_count", new Rate());
+    allowedUnintentionallyKPS = registerSensor("quota_unintentionally_allowed_key_count", new Count());
+    usageRatioSensor =
+        registerSensor(new AsyncGauge((ignored, ignored2) -> getReadQuotaUsageRatio(), "quota_requested_usage_ratio"));
+    this.time = time;
+    metricConfig = new MetricConfig(); // use default configs
+  }
+
+  public void setCurrentVersion(int version) {
```

Comment:
Can this method be invoked by multiple threads concurrently?
If yes, there is a race condition as compare/swap are not atomic here.
So if the method can only be invoked very infrequently, we can make it `synchronized`.

Reply:
Or use `AtomicInteger#compareAndSet` to use the returned `boolean` to decide whether to set `backupVersion` or not.

Reply:
Good call, I don't think it can be invoked by multiple threads concurrently as of now but just to be future-proof let me make it thread safe.